apps/science/src/main.cpp

```diff
@@ -1,19 +1,164 @@
+DigitalOut ledB(LED_B);
+
+// Declaring can object
+CAN can(CAN1_RX, CAN1_TX, HWBRIDGE::ROVERCONFIG::ROVER_CANBUS_FREQUENCY);
+
+// k interval
+const int k_interval_ms = 500;
+
+// static objects
+static mbed_error_status_t setMotionData(CANMsg &msg) {
+  float motionData;
+  msg.getPayload(motionData);
+
+  switch (msg.getID()) {
+    case HWBRIDGE::CANID::SET_GENEVA_INDEX:
+      return indexerActuator.setMotionData(motionData);
+    case HWBRIDGE::CANID::SET_SCOOPER_ANGLE:
+    case HWBRIDGE::CANID::SET_COVER_ANGLE:
+      return coverServo.setPosition(motionData);
+    case HWBRIDGE::CANID::SET_ELEVATOR_HEIGHT:
+      return diggerServo.setPosition(motionData);
+    default:
+      return MBED_ERROR_INVALID_ARGUMENT;
+  }
+}
+
+static CANMsg::CANMsgHandlerMap canHandleMap = {{HWBRIDGE::CANID::SET_GENEVA_INDEX, setMotionData},
+                                                {HWBRIDGE::CANID::SET_SCOOPER_ANGLE, setMotionData},
+                                                {HWBRIDGE::CANID::SET_COVER_ANGLE, setMotionData},
+                                                {HWBRIDGE::CANID::SET_ELEVATOR_HEIGHT, setMotionData}};
+
+// CAN Threads
+Thread rxCANProcessorThread;
+Thread txCANProcessorThread;
+
+// recieving CAN messages
+void rxCANProcessor() {
+  CANMsg rxMsg;
 
-constexpr auto k_interval = 500ms;
+  // new while loop
+  while (true) {
+    if (can.read(rxMsg)) {
+      if (canHandleMap.size() > 0) {
+        canHandleMap[rxMsg.getID()](rxMsg);
+      } else {
+        // canHandleMap has not been declared or there is some other error
+      }
+    }
+  }
+
+  /*
+  // this is the old while loop
+    while (true) {
+      if (can.read(rxMsg)) {
+        if (canHandleMap.count(rxMsg.getID()) > 0) {
+          canHandleMap[rxMsg.getID()](rxMsg);
+        } else {
+          // ruh roh
+        }
+      }
+    }
+    */
+}
+
+// Send outgoing CAN messages
+void txCANProcessor() {
+  constexpr std::chrono::milliseconds txPeriod = 500ms;
+  CANMsg txMsg;
+
+  while (true) {
+    txMsg.setID(HWBRIDGE::CANID::REPORT_GENEVA_INDEX);
+    txMsg.setPayload(indexerActuator.getAngle_Degrees());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+
+    txMsg.setID(HWBRIDGE::CANID::REPORT_SCOOPER_ANGLE);
+    txMsg.setPayload(elevatorActuator.getAngle_Degrees());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
 
-CANBus can(CAN1_RX, CAN1_TX, HWBRIDGE::ROVERCONFIG::ROVER_CANBUS_FREQUENCY);
+    txMsg.setID(HWBRIDGE::CANID::REPORT_COVER_ANGLE);
+    txMsg.setPayload(coverServo.read());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+
+    txMsg.setID(HWBRIDGE::CANID::REPORT_ELEVATOR_HEIGHT);
+    txMsg.setPayload(diggerServo.read());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+
+    // Read moisture returns an unsigned number so it needs to be cast to an int to be handled
+    txMsg.setID(HWBRIDGE::CANID::REPORT_MOISTURE_DATA);
+    txMsg.setPayload((int)moistureSensor.Read_Moisture());
```

Comment:
not sure what our policy is around C style casts but you could consider using static cast such as `static_cast<int>(moistureSensor.Read_Moisture())`

Reply:
im not sure either but i changed it to what you suggested